File: Data Structures & Algorithms/longest-substring-with-at-most-two-distinct-characters/submission-2.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        int longest = 0;
        int left = 0;

        HashMap<Character, Integer> mapping = new HashMap<>();

        for (int right = 0; right < s.length(); right++) {
            char rightChar = s.charAt(right);

            mapping.put(
                rightChar,
                mapping.getOrDefault(rightChar, 0) + 1
            );

            while (mapping.size() > 2) {
                char leftChar = s.charAt(left);

                mapping.put(
                    leftChar,
                    mapping.get(leftChar) - 1
                );

                if (mapping.get(leftChar) == 0) {
                    mapping.remove(leftChar);
                }

                left++;
            }

            longest = Math.max(longest, right - left + 1);
        }

        return longest;
    }
}

File: Data Structures & Algorithms/max-consecutive-ones-iii/submission-2.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        max_length = 0
        zeros_flipped = 0
        left = 0

        for right in range(len(nums)):
            if nums[right] == 0:
                zeros_flipped += 1
            
            while zeros_flipped > k:
                if nums[left] == 0:
                    zeros_flipped -= 1
                left += 1
            max_length = max(max_length, right-left+1)
        return max_length
        